<a href="https://colab.research.google.com/github/naraphattul/229352-StatisticalLearning/blob/main/%E0%B8%B8%E0%B8%B8660510769_Lab04_Naive_Bayes_Grid_and_Random_Search.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Statistical Learning for Data Science 2 (229352)
#### Instructor: Donlapark Ponnoprat

#### [Course website](https://donlapark.pages.dev/229352/)

## Lab #4

In [1]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

from scipy.stats import uniform

In [2]:
train = fetch_20newsgroups(subset='train')
test = fetch_20newsgroups(subset='test')

Xtrain = train.data[:3000]
ytrain = train.target[:3000]
Xtest = test.data[:500]
ytest = test.target[:500]

print("X:", len(Xtest))
print("y:", len(ytest))

X: 500
y: 500


### Naive Bayes [(Documentation)](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html)

In [3]:
from sklearn.naive_bayes import MultinomialNB

nb = MultinomialNB()

### Random Search Cross-Validation [(Documentation)](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html)

### Uniform distribution in `Scipy` [(Documentation)](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.uniform.html)

In [4]:
from inspect import Parameter
from sklearn.feature_extraction.text import CountVectorizer

pipeline = Pipeline([
    ('count', CountVectorizer(stop_words='english')),
    ('nb', MultinomialNB())])

parameters = {'nb__alpha': uniform(loc=0.1, scale=10)}

clf = RandomizedSearchCV(pipeline, parameters, n_iter=10, cv=3)
clf.fit(Xtrain, ytrain)

RandomizedSearchCV(cv=3,
                   estimator=Pipeline(steps=[('count',
                                              CountVectorizer(stop_words='english')),
                                             ('nb', MultinomialNB())]),
                   param_distributions={'nb__alpha': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7c30aa248750>})

In [5]:
ypred = clf.predict(Xtest)
print(classification_report(ytest, ypred))

              precision    recall  f1-score   support

           0       0.67      0.67      0.67        21
           1       0.80      0.38      0.52        21
           2       1.00      0.08      0.14        26
           3       0.48      0.79      0.60        34
           4       0.92      0.68      0.78        34
           5       0.57      0.81      0.67        26
           6       1.00      0.73      0.84        22
           7       0.75      0.96      0.84        28
           8       1.00      0.76      0.86        33
           9       0.88      0.84      0.86        25
          10       0.84      1.00      0.92        27
          11       0.54      1.00      0.70        20
          12       0.67      0.42      0.51        24
          13       0.86      0.83      0.84        23
          14       0.81      0.89      0.85        28
          15       0.71      0.93      0.81        29
          16       0.51      0.95      0.67        21
          17       0.82    

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


#### Exercise

1. For the Naive Bayes model, use grid search 5-fold cross-validation across different values of `alpha` to find the best model.

2. For the best value of `alpha`, compute the `f1_macro` score on the test set.
* What value of `alpha` did you obtain?
* What is the model's `f1_macro` score?

3. Repeat Exercise 1 and 2 for **random search** 5-fold cross validation across different values of `alpha`. Compute the `f1_macro` score on the test set.
* What value of `alpha` did you obtain?
* Did you get a better `f1_macro` score compared to grid search in Exercise 2?

In [6]:
# Exercise 1 Used Grid Search 5-fold Cross-Validation
from sklearn.metrics import f1_score

train = fetch_20newsgroups(subset='train')
test = fetch_20newsgroups(subset='test')

Xtrain = train.data[:3000]
ytrain = train.target[:3000]
Xtest = test.data[:500]
ytest = test.target[:500]

pipeline_grid = Pipeline([
    ('count', CountVectorizer(stop_words='english')),
    ('nb', MultinomialNB())])

alpha_values = [0.01, 0.1, 0.5, 1.0, 2.0, 5.0, 10.0]
param_grid = {'nb__alpha': alpha_values}

grid_search = GridSearchCV(
    pipeline_grid,
    param_grid,
    cv=5,
    scoring='f1_macro',
    n_jobs=-1
)

grid_search.fit(Xtrain, ytrain)
best_alpha_grid = grid_search.best_params_['nb__alpha']

In [11]:
# Exercise 2 Best Value of ฟlpha, F1-Macro Score On The Test Set

y_pred_grid = grid_search.predict(Xtest)
f1_macro_grid = f1_score(ytest, y_pred_grid, average='macro')

print(f'Best alpha (Grid): {best_alpha_grid}')
print(f'F1-macro score (Grid): {f1_macro_grid}')

Best alpha (Grid): 0.01
F1-macro score (Grid): 0.7077459540233744


In [9]:
# Exercise 3 Used Random Serch 5-fold Cross-Validation

train = fetch_20newsgroups(subset='train')
test = fetch_20newsgroups(subset='test')

Xtrain = train.data[:3000]
ytrain = train.target[:3000]
Xtest = test.data[:500]
ytest = test.target[:500]

pipeline_random = Pipeline([
    ('count', CountVectorizer(stop_words='english')),
    ('nb', MultinomialNB())])

param_random = {'nb__alpha': uniform(loc=0.01, scale=99.99)}

random_search = RandomizedSearchCV(
    pipeline_random,
    param_random,
    n_iter=20, # random sample
    cv=5,
    scoring='f1_macro',
    random_state=42,
    n_jobs=-1)

random_search.fit(Xtrain, ytrain)
best_alpha_random = random_search.best_params_['nb__alpha']
print(f'Best alpha (Random): {best_alpha_random}')

y_pred_random = random_search.predict(Xtest)
f1_macro_random = f1_score(ytest, y_pred_random, average='macro')
print(f'F1-macro score (Random): {f1_macro_random}')

Best alpha (Random): 2.0682435846372864
F1-macro score (Random): 0.6620170789764999


In [10]:
ypred_random = random_search.predict(Xtest)
print(classification_report(ytest, ypred_random))

              precision    recall  f1-score   support

           0       0.68      0.62      0.65        21
           1       0.88      0.33      0.48        21
           2       1.00      0.15      0.27        26
           3       0.50      0.82      0.62        34
           4       0.91      0.59      0.71        34
           5       0.55      0.85      0.67        26
           6       1.00      0.73      0.84        22
           7       0.82      0.96      0.89        28
           8       1.00      0.73      0.84        33
           9       0.95      0.80      0.87        25
          10       0.84      1.00      0.92        27
          11       0.49      1.00      0.66        20
          12       0.82      0.38      0.51        24
          13       0.83      0.83      0.83        23
          14       0.75      0.86      0.80        28
          15       0.71      0.93      0.81        29
          16       0.44      0.95      0.61        21
          17       0.69    

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Grid Search ได้ f1_macro ที่มากกว่า Random Serch (0.7077 > 0.6620)
เพราะค่า alpha ของ Grid Search เหมาสมกับลักษณะของข้อมูลมากกว่า Random Serch

